File: backend/src/main/java/cn/etarch/mao/harness/mcp/McpToolAdapter.java
package cn.etarch.mao.harness.mcp;

import cn.etarch.mao.harness.mcp.model.McpToolRef;
import cn.etarch.mao.harness.tool.Tool;
import lombok.extern.slf4j.Slf4j;

import java.util.Map;

/**
 * 将 MCP 工具包装为标准 {@link Tool} 接口，使 harness 全链路（schema 注入、执行调度、
 * 结果摘要）透明复用。
 * <p>
 * 命名：{@code mcp__{serverName}__{toolName}}，与内置工具命名空间隔离。
 * <p>
 * 执行语义：
 * <ul>
 *   <li>CLOUD 模式：{@link McpClientManager#callTool} 直连执行（ToolDispatcher 直接调用本 adapter）；</li>
 *   <li>LOCAL 模式：本 adapter 仅承载 schema；执行由 ToolDispatcher 的 LOCAL 分支按完整工具名
 *       委托桌面端（Electron）经 WebSocket 执行，桌面端按 {@code serverName + toolName} 路由。</li>
 * </ul>
 */
@Slf4j
public class McpToolAdapter implements Tool {

    private final McpToolRef ref;
    private final McpClientManager clientManager;

    public McpToolAdapter(McpToolRef ref, McpClientManager clientManager) {
        this.ref = ref;
        this.clientManager = clientManager;
    }

    @Override
    public String getName() {
        return ref.getFullToolName();
    }

    @Override
    public String getDescription() {
        return ref.getDescription();
    }

    @Override
    public Map<String, Object> getInputSchema() {
        return ref.getInputSchema();
    }

    @Override
    public Map<String, Object> getOutputSchema() {
        return Map.of("type", "object", "description", "MCP 工具执行结果");
    }

    @Override
    public String execute(String arguments) {
        return execute(arguments, null, null, null);
    }

    @Override
    public String execute(String arguments, String workspace) {
        return execute(arguments, null, null, workspace);
    }

    @Override
    public String execute(String arguments, Long sessionId, String workspace) {
        return execute(arguments, sessionId, null, workspace);
    }

    @Override
    public String execute(String arguments, Long sessionId, Long userId, String workspace) {
        if (clientManager == null) {
            return "{\"error\":\"MCP 工具在 LOCAL 模式下由桌面端执行，服务端无法直接调用\"}";
        }
        return clientManager.callTool(sessionId, ref.getServerId(), ref.getToolName(), arguments);
    }

    public McpToolRef getRef() {
        return ref;
    }
}

File: backend-ts/src/harness/mcp/mcp-tool-adapter.ts
import type { Tool } from '../tool/tool.js';
import type { ToolDescriptor } from '../tool/tool-descriptor.js';
import { fullToolName, normalizeMcpInputSchema, type McpToolRef } from './entity/mcp-server.js';
import type { McpClientManager } from './mcp-client-manager.js';

export class McpToolAdapter implements Tool {
  constructor(
    private readonly ref: McpToolRef,
    private readonly clientManager: McpClientManager | null,
  ) {}

  getName(): string {
    return this.ref.fullToolName ?? fullToolName(this.ref.serverName, this.ref.toolName);
  }

  getDescriptor(): ToolDescriptor {
    // clientManager 为 null 即 LOCAL 模式（服务端无 MCP 连接，由桌面端执行）——隐式约定在此显式化
    return {
      name: this.getName(),
      source: 'mcp',
      executor: this.clientManager ? 'mcp-server' : 'desktop',
      serverId: this.ref.serverId,
      originalName: this.ref.toolName,
    };
  }

  getDescription(): string {
    return this.ref.description;
  }

  getInputSchema(): Record<string, unknown> {
    const raw = this.ref.inputSchema
      ?? (this.ref as { schema?: Record<string, unknown> }).schema
      ?? {};
    return normalizeMcpInputSchema(raw);
  }

  getOutputSchema(): Record<string, unknown> {
    return { type: 'object', description: 'MCP 工具执行结果' };
  }

  execute(argumentsJson: string, a?: unknown, _b?: unknown, _c?: unknown): string | Promise<string> {
    if (!this.clientManager) {
      return '{"error":"MCP 工具在 LOCAL 模式下由桌面端执行，服务端无法直接调用"}';
    }
    const sessionId = typeof a === 'number' ? a : null;
    return this.clientManager.callTool(sessionId, this.ref.serverId, this.ref.toolName, argumentsJson);
  }

  getRef(): McpToolRef {
    return this.ref;
  }
}
